plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi_test.go

```diff
@@ -0,0 +1,239 @@
+
+	parsed, err = parsePath("", "", "")
+	assert.Nil(t, err)
+	assert.Equal(t, *parsed, gnmi.Path{})
+
+	parsed, err = parsePath("", "/foo[[", "")
+	assert.Nil(t, parsed)
+	assert.Equal(t, errors.New("Invalid GNMI path: /foo[[/"), err)
+}
+
+type mockGNMIServer struct {
+	t        *testing.T
+	server   *grpc.Server
+	scenario int
+}
+
+func (m *mockGNMIServer) Capabilities(context.Context, *gnmi.CapabilityRequest) (*gnmi.CapabilityResponse, error) {
+	return nil, nil
+}
+
+func (m *mockGNMIServer) Get(context.Context, *gnmi.GetRequest) (*gnmi.GetResponse, error) {
+	return nil, nil
+}
+
+func (m *mockGNMIServer) Set(context.Context, *gnmi.SetRequest) (*gnmi.SetResponse, error) {
+	return nil, nil
+}
+
+func (m *mockGNMIServer) Subscribe(server gnmi.GNMI_SubscribeServer) error {
+	defer func() {
+		if m.scenario >= 0 {
+			m.scenario = -1
+			time.AfterFunc(250*time.Millisecond, m.server.Stop)
+		}
+	}()
+
+	metadata, ok := metadata.FromIncomingContext(server.Context())
+	assert.Equal(m.t, ok, true)
+	assert.Equal(m.t, metadata.Get("username"), []string{"theuser"})
+	assert.Equal(m.t, metadata.Get("password"), []string{"thepassword"})
+
+	switch m.scenario {
+	case 0:
+		return fmt.Errorf("testerror")
+	case 1:
+		notification := mockGNMINotification()
+		server.Send(&gnmi.SubscribeResponse{Response: &gnmi.SubscribeResponse_Update{Update: notification}})
+		server.Send(&gnmi.SubscribeResponse{Response: &gnmi.SubscribeResponse_SyncResponse{SyncResponse: true}})
+		notification.Prefix.Elem[0].Key["foo"] = "bar2"
+		notification.Update[0].Path.Elem[1].Key["name"] = "str2"
+		notification.Update[0].Val = &gnmi.TypedValue{Value: &gnmi.TypedValue_JsonVal{JsonVal: []byte{'"', '1', '2', '3', '"'}}}
+		server.Send(&gnmi.SubscribeResponse{Response: &gnmi.SubscribeResponse_Update{Update: notification}})
+		return nil
+	case 2:
+		notification := mockGNMINotification()
+		server.Send(&gnmi.SubscribeResponse{Response: &gnmi.SubscribeResponse_Update{Update: notification}})
+		return nil
+	case 3:
+		notification := mockGNMINotification()
+		notification.Prefix.Elem[0].Key["foo"] = "bar2"
+		notification.Update[0].Path.Elem[1].Key["name"] = "str2"
+		notification.Update[0].Val = &gnmi.TypedValue{Value: &gnmi.TypedValue_BoolVal{BoolVal: false}}
+		server.Send(&gnmi.SubscribeResponse{Response: &gnmi.SubscribeResponse_Update{Update: notification}})
+		return nil
+	default:
+		return fmt.Errorf("test not implemented ;)")
+	}
+}
+
+func TestGNMIError(t *testing.T) {
+	listener, _ := net.Listen("tcp", "127.0.0.1:57003")
+	server := grpc.NewServer()
+	gnmi.RegisterGNMIServer(server, &mockGNMIServer{t: t, scenario: 0, server: server})
+
+	c := &CiscoTelemetryGNMI{Addresses: []string{"127.0.0.1:57003"},
+		Username: "theuser", Password: "thepassword", Encoding: "proto",
+		Redial: internal.Duration{Duration: 1 * time.Second}}
+
+	acc := &testutil.Accumulator{}
+	assert.Nil(t, c.Start(acc))
+	server.Serve(listener)
+	c.Stop()
```

Comment:
To avoid test race, call acc.WaitError(n) before calling Stop

Reply:
Done, thanks!